linker/validate.go

```diff
@@ -25,29 +25,36 @@ import (
 
 	"github.com/bufbuild/protocompile/internal"
 	"github.com/bufbuild/protocompile/reporter"
+	"github.com/bufbuild/protocompile/walk"
 )
 
 // ValidateOptions runs some validation checks on the result that can only
 // be done after options are interpreted.
 func (r *result) ValidateOptions(handler *reporter.Handler) error {
-	if err := r.validateExtensions(r, handler); err != nil {
-		return err
-	}
-	return r.validateJSONNamesInFile(handler)
```

Comment:
When trying to figure out the best way to incorporate the packed option check into these existing traversals of the hierarchy, I figured it would be better to get rid of the manual recursive traversal and do this more like a visitor pattern. The existing `walk` package in this repo was intended exactly for this sort of thing.